### PR TITLE
Feat/git update

### LINKED ISSUE #65: Read licence in git scraper
Use the git repo scraper to read and categorize the license

### PR DESCRIPTION
**Pull request by karamba228**

### CODE REVIEW

File: score/git_vcs/scrape.py [hunk 1]
```diff
@@ -14,25 +15,22 @@
...
         try:
             repo = Repo.clone_from(url, tmpdir, no_checkout=True, filter="tree:0")
-        except UnsafeProtocolError:
-            return {"error": "Unsafe Git Protocol: protocol looks suspicious"}
-        except GitCommandError as err:
-            if err.status == 128 and "not found" in err.stderr:
-                return {"error": "Repo not found"}
-            log.error(f"{url}: {err.stderr}")
+        except (UnsafeProtocolError, GitCommandError) as err:
+            log.error(f"{url}: {err}")
```
Comment (srossross) [inline score/git_vcs/scrape.py]: why did you remove the 'Repo not found' error?

File: score/git_vcs/scrape.py [hunk 2]
```diff
@@ -68,3 +59,32 @@ def get_info_from_git_repo(url) -> dict:
...
         }
+
+
+def get_license_type(url) -> str:
+    with tempfile.TemporaryDirectory(
+        prefix="score", suffix=".git", ignore_cleanup_errors=True
+    ) as tmpdir:
+        try:
+            # Perform a shallow clone with depth=1 to fetch the LICENSE file
+            repo = Repo.clone_from(url, tmpdir, depth=1)
```
Comment (srossross) [inline score/git_vcs/scrape.py]: this is cloning the repo again right? can you move this so we don't clone the repo twice?

**Review by srossross — CHANGES_REQUESTED** [review]

File: score/git_vcs/scrape.py (hunk 2 above)
Comment (karamba228) [reply score/git_vcs/scrape.py]: This is the fastest way to do this task. For the commit data we need to go full depth of the repository, so we do `no_checkout` for the fastest result. On the other hand to get the `License ` we need to check out the repo so we can read the file, but we can do `depth=1`  to speed it up. Doing 2 requests separately ends up being faster than doing a full depth checkout of some repositories.

I tested it with 13 different projects that had different types of licenses to make sure the detection is accurate.
#### Here are some time results I got:
* Without license detection (just retrieving the commit history): `3min 10s`
* With license detection but request separate: `7min 15s`
* With one request but full depth checkout: It had 6 projects processed at `7min 40s`

Comment (srossross) [reply score/git_vcs/scrape.py]: Thanks for the explanation:

On the bare repo can you just run:

```py
repo.git.checkout(repo.active_branch, "--", "LICENSE*")
```

this should get the licence and only check out `LICENSE.txt` or `LICENSE`

File: score/git_vcs/scrape.py [hunk 3]
```diff
@@ -45,26 +53,64 @@ def get_info_from_git_repo(url) -> dict:
...
+                log.error(f"{url}: Repo not found")
+                return "No License Found"
+            log.error(f"{url}: {err.stderr}")
+            return "No License Found"
+
+        # Check for a LICENSE file in the root directory
+        license_content = None
+        try:
+            root_files = repo.tree().blobs  # List blobs (files) in the repository root
+            for blob in root_files:
```
Comment (srossross) [inline score/git_vcs/scrape.py]: Please do no not iterate over all blobs, just check for the existence of `LICENSE` or `LICENSE.txt` in the root directory with `os.path.exists`

File: score/git_vcs/license_detection.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+import re
+
+
+def identify_license(license_content: str) -> str:
```
Comment (srossross) [inline score/git_vcs/license_detection.py:4]: This https://github.com/licensee/licensee seems like a good ruby implementation of a licence checker, can you look into how to implement this in python?

**Review by srossross — APPROVED** [review]

**Pull request merged**